// dynamic-mat-table/esm2015/lib/dynamic-mat-table/extensions/row-menu/row-menu.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-mat-table/extensions/row-menu/row-menu.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Output, Input, EventEmitter } from '@angular/core';
import { isNullorUndefined } from '../../../cores/type';
/**
 * @template T
 */
export class RowMenuComponent {
    constructor() {
        // @Output() rowActionChange: EventEmitter<ContextMenuItem> = new EventEmitter<ContextMenuItem>();
        this.rowActionChange = new EventEmitter();
        this.actionMenus = [];
        this.visibleActionMenus = [];
    }
    /**
     * @param {?} e
     * @return {?}
     */
    menuOnClick(e) {
        e.stopPropagation();
        e.preventDefault();
        this.visibleActionMenus = [];
        this.actionMenus.forEach((/**
         * @param {?} menu
         * @return {?}
         */
        menu => {
            /** @type {?} */
            const am = isNullorUndefined(this.rowActionMenu) || isNullorUndefined(this.rowActionMenu[menu.name]) ? menu : this.rowActionMenu[menu.name];
            if (isNullorUndefined(am.visible) || am.visible) {
                this.visibleActionMenus.push({
                    name: menu.name,
                    text: am.text || menu.text,
                    disabled: am.disabled || menu.disabled,
                    icon: am.icon || menu.icon,
                    color: am.color || menu.color
                });
            }
        }));
    }
    /**
     * @param {?} menu
     * @return {?}
     */
    menuButton_OnClick(menu) {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.rowActionChange.emit(menu);
        }));
    }
}
RowMenuComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line: component-selector
                selector: 'row-menu',
                template: "<button class=\"clear\" mat-icon-button #menuTrigger=\"matMenuTrigger\" (click)=\"menuOnClick($event)\"\r\n[matMenuTriggerFor]=\"menu\" [dir]=\"tableSetting.direction === 'rtl' ? 'ltr' : 'rtl'\" >\r\n  <mat-icon>more_horiz</mat-icon>\r\n</button>\r\n\r\n<mat-menu #menu=\"matMenu\" [overlapTrigger]=\"false\" [dir]=\"tableSetting.direction === 'rtl' ? 'ltr' : 'rtl'\">\r\n  <ng-template matMenuContent>\r\n    <button mat-button [class.ltr-menu]=\"tableSetting.direction === 'rtl'\" [color]=\"menu.color\" class=\"button-menu\" \r\n      *ngFor=\"let menu of visibleActionMenus\" [disabled]=\"menu.disabled\" (click)=\"menuButton_OnClick(menu)\">\r\n       <mat-icon>{{menu.icon}}</mat-icon>\r\n      <span [class.text-align-right]=\"tableSetting.direction === 'rtl'\" class=\"text-align-left\">{{menu.text}}</span>\r\n    </button>\r\n  </ng-template>\r\n</mat-menu>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["@media print{.printpreview{background-color:#fff;position:fixed;width:100%;height:auto;z-index:99999999;margin:0;padding:0;top:0;left:0;overflow:visible;display:block}}.disable-backdrop-click .cdk-overlay-backdrop.cdk-overlay-transparent-backdrop.cdk-overlay-backdrop-showing{pointer-events:none}.ltr-menu span{float:left}.button-menu{width:100%;line-height:48px}.button-menu::ng-deep .mat-button-wrapper{display:flex}.button-menu::ng-deep .mat-button-wrapper span{display:inline-block;width:100%;text-align:left}.button-menu::ng-deep .mat-button-wrapper mat-icon{line-height:48px;height:48px;margin:0 5px}mat-button-wrapper .button-menu{display:inline-block!important}.text-align-left{text-align:left!important}.text-align-right{text-align:right!important}.mat-menu-panel{min-height:unset!important}:host{display:flex;align-items:center;justify-content:space-between}"]
            }] }
];
/** @nocollapse */
RowMenuComponent.ctorParameters = () => [];
RowMenuComponent.propDecorators = {
    rowActionChange: [{ type: Output }],
    actionMenus: [{ type: Input }],
    tableSetting: [{ type: Input }],
    rowActionMenu: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    RowMenuComponent.prototype.rowActionChange;
    /** @type {?} */
    RowMenuComponent.prototype.actionMenus;
    /** @type {?} */
    RowMenuComponent.prototype.tableSetting;
    /** @type {?} */
    RowMenuComponent.prototype.rowActionMenu;
    /** @type {?} */
    RowMenuComponent.prototype.visibleActionMenus;
}
