// dynamic-mat-table/esm5/lib/dynamic-mat-table/extensions/filter/compare/abstract-filter.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-mat-table/extensions/filter/compare/abstract-filter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 * @template T
 */
var /**
 * @abstract
 * @template T
 */
AbstractFilter = /** @class */ (function () {
    function AbstractFilter() {
    }
    /**
     * @return {?}
     */
    AbstractFilter.prototype.hasValue = /**
     * @return {?}
     */
    function () {
        if (this.parameters !== null) {
            return this.parameters.filter((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return p.value != null && p.value !== undefined && p.value.toString() !== ''; })).length > 0;
        }
    };
    return AbstractFilter;
}());
/**
 * @abstract
 * @template T
 */
export { AbstractFilter };
if (false) {
    /** @type {?} */
    AbstractFilter.prototype.parameters;
    /** @type {?} */
    AbstractFilter.prototype.type;
    /** @type {?} */
    AbstractFilter.prototype.selectedIndex;
    /** @type {?} */
    AbstractFilter.prototype.selectedValue;
    /**
     * @abstract
     * @param {?} dynamicVariable
     * @return {?}
     */
    AbstractFilter.prototype.toString = function (dynamicVariable) { };
    /**
     * @abstract
     * @return {?}
     */
    AbstractFilter.prototype.toPrint = function () { };
    /**
     * @abstract
     * @return {?}
     */
    AbstractFilter.prototype.toSql = function () { };
    /**
     * @abstract
     * @return {?}
     */
    AbstractFilter.prototype.getOperations = function () { };
}
/**
 * @record
 */
export function FilterOperation() { }
if (false) {
    /** @type {?} */
    FilterOperation.prototype.predicate;
    /** @type {?} */
    FilterOperation.prototype.text;
}
