// dynamic-mat-table/esm5/lib/dynamic-mat-table/extensions/filter/compare/text-filter.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-mat-table/extensions/filter/compare/text-filter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __extends } from "tslib";
import { AbstractFilter } from './abstract-filter';
/** @type {?} */
var contains = 'a.toString().includes(b)';
/** @type {?} */
var equals = 'a.toString() === b.toString()';
/** @type {?} */
var startsWith = 'a.toString().startsWith(b)';
/** @type {?} */
var endsWith = 'a.toString().endsWith(b.toString())';
/** @type {?} */
var empty = '!a';
/** @type {?} */
var notEmpty = '!!a';
/** @type {?} */
var operations = [contains, equals, startsWith, endsWith, empty, notEmpty];
var TextFilter = /** @class */ (function (_super) {
    __extends(TextFilter, _super);
    function TextFilter(languagePack) {
        var _this = _super.call(this) || this;
        _this.languagePack = languagePack;
        // tslint:disable-next-line:variable-name
        _this._selectedIndex = null;
        _this._selectedIndex = 0;
        if (TextFilter.operationList.length === 0) { // init for first time
            operations.forEach((/**
             * @param {?} fn
             * @return {?}
             */
            function (fn) {
                TextFilter.operationList.push({ predicate: fn, text: null });
            }));
        }
        TextFilter.operationList[0].text = languagePack.filterLabels.TextContains; // contains //
        TextFilter.operationList[1].text = languagePack.filterLabels.TextEquals; // equals //
        TextFilter.operationList[2].text = languagePack.filterLabels.TextStartsWith; // startsWith //
        TextFilter.operationList[3].text = languagePack.filterLabels.TextEndsWith; // endsWith //
        TextFilter.operationList[4].text = languagePack.filterLabels.TextEmpty; // empty //
        TextFilter.operationList[5].text = languagePack.filterLabels.TextNotEmpty; // notEmpty //
        return _this;
    }
    Object.defineProperty(TextFilter.prototype, "selectedIndex", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selectedIndex;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._selectedIndex = value;
            // init filter parameters
            if (value === 0 || value === 1 || value === 2 || value === 3) { // contains equals startsWith endsWith
                this.parameters = [{ value: '', text: this.languagePack.filterLabels.Text }];
            }
            else { // empty notEmpty
                this.parameters = null;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextFilter.prototype, "selectedValue", {
        get: /**
         * @return {?}
         */
        function () {
            if (this._selectedIndex !== null) {
                return TextFilter.operationList[this._selectedIndex];
            }
            else {
                return null;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TextFilter.prototype.getOperations = /**
     * @return {?}
     */
    function () {
        return TextFilter.operationList;
    };
    /**
     * @param {?} dynamicVariable
     * @return {?}
     */
    TextFilter.prototype.toString = /**
     * @param {?} dynamicVariable
     * @return {?}
     */
    function (dynamicVariable) {
        /** @type {?} */
        var a = '_a$';
        /** @type {?} */
        var b = '_b$';
        /** @type {?} */
        var predicate = this.selectedValue.predicate.replace('a', a).replace('b', b);
        /** @type {?} */
        var statement = predicate.replace(a, a + "['" + dynamicVariable + "']?.toString()?.toLowerCase()");
        // one static parameters equals  notEquals greaterThan lessThan //
        if (this._selectedIndex === 0 ||
            this._selectedIndex === 1 ||
            this._selectedIndex === 2 ||
            this._selectedIndex === 3) {
            /** @type {?} */
            var value = '\'' + (this.parameters[0].value !== null ? this.parameters[0].value.toLowerCase() : ' null ') + '\'';
            return statement.replace('_b$', value);
        }
        else { // without static parameters
            return statement;
        }
    };
    /**
     * @return {?}
     */
    TextFilter.prototype.toPrint = /**
     * @return {?}
     */
    function () {
        return TextFilter.operationList[this._selectedIndex].text + ' ' + this.parameters[0].value + ' ' + (this.type || '') + ' ';
    };
    /**
     * @return {?}
     */
    TextFilter.prototype.toSql = /**
     * @return {?}
     */
    function () {
        return TextFilter.sql[this._selectedIndex].replace('[*]', (this.parameters[0].value || '')) + (this.type || '') + ' ';
    };
    TextFilter.sql = ['LIKE "%[*]%"', '= "[*]"', 'LIKE "%[*]"', 'LIKE "[*]%"', 'IS NULL', 'IS NOT NULL'];
    TextFilter.operationList = [];
    return TextFilter;
}(AbstractFilter));
export { TextFilter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TextFilter.sql;
    /**
     * @type {?}
     * @private
     */
    TextFilter.operationList;
    /**
     * @type {?}
     * @private
     */
    TextFilter.prototype._selectedIndex;
    /** @type {?} */
    TextFilter.prototype.languagePack;
}
