// dynamic-mat-table/esm2015/lib/cores/dynamic-cell/dynamic-cell.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/cores/dynamic-cell/dynamic-cell.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Compiler, ComponentFactoryResolver, Directive, EventEmitter, Input, ViewContainerRef } from '@angular/core';
import { DynamicMatTableComponent } from '../../dynamic-mat-table/dynamic-mat-table.component';
export class DynamicCellDirective {
    /**
     * @param {?} compiler
     * @param {?} cfr
     * @param {?} vc
     * @param {?} parent
     */
    constructor(compiler, cfr, vc, parent) {
        this.compiler = compiler;
        this.cfr = cfr;
        this.vc = vc;
        this.parent = parent;
        this.componentRef = null;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (this.componentRef === null || this.componentRef === undefined) {
            this.initComponent();
        }
        // pass input parameters
        if (changes.column && changes.column.currentValue) {
            this.componentRef.instance.column = this.column;
        }
        if (changes.row && changes.row.currentValue) {
            ((/** @type {?} */ (this.componentRef.instance))).row = this.row;
        }
        if (changes.onRowEvent && changes.onRowEvent.currentValue) {
            ((/** @type {?} */ (this.componentRef.instance))).onRowEvent = this.onRowEvent;
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.componentRef) {
            this.componentRef.destroy();
        }
    }
    /**
     * @return {?}
     */
    initComponent() {
        try {
            /** @type {?} */
            const componentFactory = this.cfr.resolveComponentFactory(this.component);
            this.componentRef = this.vc.createComponent(componentFactory);
            this.updateInput();
        }
        catch (e) {
            // console.log(e);
        }
    }
    /**
     * @return {?}
     */
    updateInput() {
        if (this.parent) {
            ((/** @type {?} */ (this.componentRef.instance))).parent = this.parent;
        }
        if (this.column) {
            this.componentRef.instance.column = this.column;
        }
        if (this.row) {
            ((/** @type {?} */ (this.componentRef.instance))).row = this.row;
        }
        if (this.onRowEvent) {
            ((/** @type {?} */ (this.componentRef.instance))).onRowEvent = this.onRowEvent;
        }
    }
}
DynamicCellDirective.decorators = [
    { type: Directive, args: [{
                selector: '[dynamicCell]'
            },] }
];
/** @nocollapse */
DynamicCellDirective.ctorParameters = () => [
    { type: Compiler },
    { type: ComponentFactoryResolver },
    { type: ViewContainerRef },
    { type: DynamicMatTableComponent }
];
DynamicCellDirective.propDecorators = {
    component: [{ type: Input }],
    column: [{ type: Input }],
    row: [{ type: Input }],
    onRowEvent: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DynamicCellDirective.prototype.component;
    /** @type {?} */
    DynamicCellDirective.prototype.column;
    /** @type {?} */
    DynamicCellDirective.prototype.row;
    /** @type {?} */
    DynamicCellDirective.prototype.onRowEvent;
    /** @type {?} */
    DynamicCellDirective.prototype.componentRef;
    /** @type {?} */
    DynamicCellDirective.prototype.compiler;
    /**
     * @type {?}
     * @private
     */
    DynamicCellDirective.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    DynamicCellDirective.prototype.vc;
    /**
     * @type {?}
     * @private
     */
    DynamicCellDirective.prototype.parent;
}
