// dynamic-mat-table/esm5/lib/dynamic-mat-table/extensions/print-dialog/print-dialog.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-mat-table/extensions/print-dialog/print-dialog.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, ElementRef, Inject } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material/dialog';
// tslint:disable-next-line:max-line-length
/** @type {?} */
var styles = 'body{margin:15px;}table{width:100%;border-collapse:collapse;}h2{text-align:center;}th.mat-header-cell{text-align:center;}div{text-align:center;margin:30px }tr{border-bottom:1px solid }td,th{padding:10px; text-align: center }.param-list{text-align: left;border:solid gray;border-width: 0px 0px 2px 0;margin-bottom: 10px;padding-bottom: 10px;}.param {display: inline-block;margin: 10px;}';
var PrintTableDialogComponent = /** @class */ (function () {
    function PrintTableDialogComponent(dialogRef, printTable) {
        this.dialogRef = dialogRef;
        this.printTable = printTable;
    }
    /**
     * @return {?}
     */
    PrintTableDialogComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    PrintTableDialogComponent.prototype.print = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var dialogConfig = 'width=600,height=700,scrollbars=no,menubar=no,toolbar=no,location=no,status=no,titlebar=no';
            /** @type {?} */
            var printDoc = "\n    <html>\n      <head>\n        <style> " + styles + " </style>\n      </head>\n      <body onload=\"window.print();\" onafterprint=\"window.close()\">\n        " + _this.printContentRef.nativeElement.innerHTML + "\n      </body>\n    </html>\n    ";
            /** @type {?} */
            var popupWinindow = window.open('', '_blank', dialogConfig);
            popupWinindow.document.write(printDoc);
            popupWinindow.document.close();
        }));
    };
    PrintTableDialogComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line: component-selector
                    selector: 'print-dialog',
                    template: "<mat-dialog-content>\r\n<div dir='{{ printTable.direction }}' #printContentRef>\r\n  <h2>\r\n      {{ printTable?.title }}\r\n  </h2>\r\n  <div class=\"param-list\">\r\n      <div class=\"param\" *ngFor='let param of printTable?.userPrintParameters'>\r\n         <b>{{ param.key }} </b> : {{ param.value }}\r\n      </div>\r\n      <div class=\"param\" *ngFor='let param of printTable?.tablePrintParameters'>\r\n        <b>{{ param.key }} </b> : {{ param.value }}\r\n     </div>\r\n  </div>\r\n  <table class=\"print-table\" mat-table [dataSource]=\"printTable.data\">\r\n      <ng-container *ngFor=\"let column of printTable.columns\" matColumnDef=\"{{ column.name }}\">\r\n        <th mat-header-cell *matHeaderCellDef> {{ column.header }} </th>\r\n        <td mat-cell *matCellDef=\"let row\"> {{ row[column.name] }} </td>\r\n      </ng-container>\r\n      <tr mat-header-row *matHeaderRowDef=\"printTable.displayedFields\"></tr>\r\n      <tr mat-row *matRowDef=\"let row; columns: printTable.displayedFields;\"></tr>\r\n  </table>\r\n</div>\r\n\r\n</mat-dialog-content>\r\n\r\n<mat-dialog-actions align=\"end\">\r\n  <button mat-button mat-dialog-close>Cancel</button>\r\n  <button mat-button [mat-dialog-close]=\"true\" cdkFocusInitial (click)=\"print()\">Print</button>\r\n</mat-dialog-actions>\r\n\r\n",
                    styles: ["#print-section{text-align:center;margin:30px}h2{text-align:center}.param-list{width:100%;display:inline-block;border:solid gray;border-width:0 0 2px;margin-bottom:10px;padding-bottom:10px}.param{display:inline-block;margin:10px}.print-table{width:100%}.print-table th.mat-header-cell{font-size:initial}"]
                }] }
    ];
    /** @nocollapse */
    PrintTableDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    PrintTableDialogComponent.propDecorators = {
        printContentRef: [{ type: ViewChild, args: ['printContentRef', { static: true },] }]
    };
    return PrintTableDialogComponent;
}());
export { PrintTableDialogComponent };
if (false) {
    /** @type {?} */
    PrintTableDialogComponent.prototype.printContentRef;
    /** @type {?} */
    PrintTableDialogComponent.prototype.dialogRef;
    /** @type {?} */
    PrintTableDialogComponent.prototype.printTable;
}
