// dynamic-mat-table/esm2015/lib/dynamic-mat-table/extensions/filter/compare/abstract-filter.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-mat-table/extensions/filter/compare/abstract-filter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 * @template T
 */
export class AbstractFilter {
    /**
     * @return {?}
     */
    hasValue() {
        if (this.parameters !== null) {
            return this.parameters.filter((/**
             * @param {?} p
             * @return {?}
             */
            p => p.value != null && p.value !== undefined && p.value.toString() !== '')).length > 0;
        }
    }
}
if (false) {
    /** @type {?} */
    AbstractFilter.prototype.parameters;
    /** @type {?} */
    AbstractFilter.prototype.type;
    /** @type {?} */
    AbstractFilter.prototype.selectedIndex;
    /** @type {?} */
    AbstractFilter.prototype.selectedValue;
    /**
     * @abstract
     * @param {?} dynamicVariable
     * @return {?}
     */
    AbstractFilter.prototype.toString = function (dynamicVariable) { };
    /**
     * @abstract
     * @return {?}
     */
    AbstractFilter.prototype.toPrint = function () { };
    /**
     * @abstract
     * @return {?}
     */
    AbstractFilter.prototype.toSql = function () { };
    /**
     * @abstract
     * @return {?}
     */
    AbstractFilter.prototype.getOperations = function () { };
}
/**
 * @record
 */
export function FilterOperation() { }
if (false) {
    /** @type {?} */
    FilterOperation.prototype.predicate;
    /** @type {?} */
    FilterOperation.prototype.text;
}
