// dynamic-mat-table/esm5/lib/cores/fixed-size-table-virtual-scroll-strategy.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/cores/fixed-size-table-virtual-scroll-strategy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { distinctUntilChanged } from 'rxjs/operators';
import { BehaviorSubject, Subject } from 'rxjs';
/**
 * @record
 */
export function TSVStrategyConfigs() { }
if (false) {
    /** @type {?} */
    TSVStrategyConfigs.prototype.rowHeight;
    /** @type {?} */
    TSVStrategyConfigs.prototype.headerHeight;
    /** @type {?} */
    TSVStrategyConfigs.prototype.footerHeight;
    /** @type {?} */
    TSVStrategyConfigs.prototype.bufferMultiplier;
}
var FixedSizeTableVirtualScrollStrategy = /** @class */ (function () {
    function FixedSizeTableVirtualScrollStrategy() {
        this.length = 0;
        this.indexChange = new Subject();
        this.stickyChange = new Subject();
        this.scrollStrategyMode = 'fixed-size';
        this.renderedRangeStream = new BehaviorSubject({ start: 0, end: 0 });
        this.offsetChange = new BehaviorSubject(0);
        this.scrolledIndexChange = this.indexChange.pipe(distinctUntilChanged());
    }
    Object.defineProperty(FixedSizeTableVirtualScrollStrategy.prototype, "dataLength", {
        get: /**
         * @return {?}
         */
        function () {
            return this.length;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.length = value;
            this.onDataLengthChanged();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.eventsSubscription.unsubscribe();
    };
    /**
     * @param {?} viewport
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.attach = /**
     * @param {?} viewport
     * @return {?}
     */
    function (viewport) {
        this.viewport = viewport;
        this.eventsSubscription = this.viewport.renderedRangeStream.subscribe(this.renderedRangeStream);
        this.onDataLengthChanged();
    };
    /**
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.detach = /**
     * @return {?}
     */
    function () {
        this.indexChange.complete();
        this.stickyChange.complete();
        this.renderedRangeStream.complete();
    };
    /**
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.onContentScrolled = /**
     * @return {?}
     */
    function () {
        this.updateContent();
    };
    /**
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.onDataLengthChanged = /**
     * @return {?}
     */
    function () {
        if (this.viewport) {
            this.viewport.setTotalContentSize(this.dataLength * this.rowHeight + this.headerHeight + this.footerHeight);
        }
        this.updateContent();
    };
    /**
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.onContentRendered = /**
     * @return {?}
     */
    function () {
        // no-op
    };
    /**
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.onRenderedOffsetChanged = /**
     * @return {?}
     */
    function () {
        // no-op
    };
    /**
     * @param {?} index
     * @param {?} behavior
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.scrollToIndex = /**
     * @param {?} index
     * @param {?} behavior
     * @return {?}
     */
    function (index, behavior) {
        // if (this.viewport) {
        //   this.viewport.scrollToOffset( this.rowHeight * index , behavior);
        // }    
        if (!this.viewport || !this.rowHeight) {
            return;
        }
        this.viewport.scrollToOffset((index - 1) * this.rowHeight + this.headerHeight);
    };
    /**
     * @param {?} configs
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.setConfig = /**
     * @param {?} configs
     * @return {?}
     */
    function (configs) {
        var rowHeight = configs.rowHeight, headerHeight = configs.headerHeight, footerHeight = configs.footerHeight, bufferMultiplier = configs.bufferMultiplier;
        if (this.rowHeight === rowHeight
            && this.headerHeight === headerHeight
            && this.footerHeight === footerHeight
            && this.bufferMultiplier === bufferMultiplier) {
            return;
        }
        this.rowHeight = rowHeight;
        this.headerHeight = headerHeight;
        this.footerHeight = footerHeight;
        this.bufferMultiplier = bufferMultiplier;
        this.onDataLengthChanged();
    };
    // bug fixed some time viewport is zero height (i dont know why!)
    // bug fixed some time viewport is zero height (i dont know why!)
    /**
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.getViewportSize = 
    // bug fixed some time viewport is zero height (i dont know why!)
    /**
     * @return {?}
     */
    function () {
        if (this.viewport.getViewportSize() === 0) {
            return this.viewport.elementRef.nativeElement.clientHeight + 52;
        }
        else {
            return this.viewport.getViewportSize();
        }
    };
    /**
     * @private
     * @return {?}
     */
    FixedSizeTableVirtualScrollStrategy.prototype.updateContent = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.viewport || !this.rowHeight) {
            return;
        }
        /** @type {?} */
        var start = 0;
        /** @type {?} */
        var end = this.dataLength;
        if (this.scrollStrategyMode === 'none' && this.viewport.getRenderedRange().start === start && this.viewport.getRenderedRange().end === end) {
            return;
        }
        /** @type {?} */
        var scrollOffset = this.viewport.measureScrollOffset();
        /** @type {?} */
        var amount = Math.ceil(this.getViewportSize() / this.rowHeight);
        /** @type {?} */
        var offset = Math.max(scrollOffset - this.headerHeight, 0);
        /** @type {?} */
        var buffer = Math.ceil(amount * this.bufferMultiplier);
        /** @type {?} */
        var skip = Math.round(offset / this.rowHeight);
        /** @type {?} */
        var index = Math.max(0, skip);
        if (this.scrollStrategyMode === 'fixed-size') {
            start = Math.max(0, index - buffer);
            end = Math.min(this.dataLength, index + amount + buffer);
        }
        /** @type {?} */
        var renderedOffset = start * this.rowHeight;
        this.viewport.setRenderedContentOffset(renderedOffset);
        this.viewport.setRenderedRange({ start: start, end: end });
        this.indexChange.next(index);
        this.stickyChange.next(renderedOffset);
        this.offsetChange.next(offset);
    };
    FixedSizeTableVirtualScrollStrategy.decorators = [
        { type: Injectable }
    ];
    return FixedSizeTableVirtualScrollStrategy;
}());
export { FixedSizeTableVirtualScrollStrategy };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FixedSizeTableVirtualScrollStrategy.prototype.eventsSubscription;
    /**
     * @type {?}
     * @private
     */
    FixedSizeTableVirtualScrollStrategy.prototype.length;
    /**
     * @type {?}
     * @private
     */
    FixedSizeTableVirtualScrollStrategy.prototype.rowHeight;
    /**
     * @type {?}
     * @private
     */
    FixedSizeTableVirtualScrollStrategy.prototype.headerHeight;
    /**
     * @type {?}
     * @private
     */
    FixedSizeTableVirtualScrollStrategy.prototype.footerHeight;
    /**
     * @type {?}
     * @private
     */
    FixedSizeTableVirtualScrollStrategy.prototype.bufferMultiplier;
    /**
     * @type {?}
     * @private
     */
    FixedSizeTableVirtualScrollStrategy.prototype.indexChange;
    /** @type {?} */
    FixedSizeTableVirtualScrollStrategy.prototype.stickyChange;
    /** @type {?} */
    FixedSizeTableVirtualScrollStrategy.prototype.scrollStrategyMode;
    /** @type {?} */
    FixedSizeTableVirtualScrollStrategy.prototype.viewport;
    /** @type {?} */
    FixedSizeTableVirtualScrollStrategy.prototype.renderedRangeStream;
    /** @type {?} */
    FixedSizeTableVirtualScrollStrategy.prototype.offsetChange;
    /** @type {?} */
    FixedSizeTableVirtualScrollStrategy.prototype.scrolledIndexChange;
}
