// dynamic-mat-table/esm5/lib/international/table-Intl.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/international/table-Intl.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
var TableIntl = /** @class */ (function () {
    function TableIntl() {
        this.menuLabels = {
            saveData: 'Save Data',
            fullScreen: 'Full Screen',
            columnSetting: 'Column Setting',
            saveTableSetting: 'Save Table Setting',
            clearFilter: 'Clear Filter',
            jsonFile: 'Json File',
            csvFile: 'CSV File',
            printTable: 'Print Table',
            filterMode: 'Filter Mode:',
            filterLocalMode: 'Local',
            filterServerMode: 'Server',
            sortMode: 'Sort Mode:',
            sortLocalMode: 'Local',
            sortServerMode: 'Server',
            printMode: 'Print Mode',
            printYesMode: 'Yes',
            printNoMode: 'No',
            pinMode: 'Pin Mode:',
            pinNoneMode: 'None',
            pinStartMode: 'Start',
            pinEndMode: 'End',
            thereIsNoColumn: 'There is no column.'
        };
        this.paginatorLabels = {
            changes: new Subject(),
            itemsPerPageLabel: 'Items per page:',
            nextPageLabel: 'Next Page:',
            previousPageLabel: 'Previous Page:',
            firstPageLabel: 'First Page:',
            lastPageLabel: 'Last Page:',
            getRangeLabel: (/**
             * @param {?} page
             * @param {?} pageSize
             * @param {?} length
             * @return {?}
             */
            function (page, pageSize, length) {
                if (length === 0 || pageSize === 0) {
                    return "0 of " + length;
                }
                length = Math.max(length, 0);
                /** @type {?} */
                var startIndex = page * pageSize;
                /** @type {?} */
                var endIndex = startIndex < length ?
                    Math.min(startIndex + pageSize, length) :
                    startIndex + pageSize;
                return startIndex + 1 + " - " + endIndex + " of " + length;
            })
        };
        this.tableLabels = {
            NoData: 'No records found.'
        };
        this.filterLabels = {
            Clear: 'Clear',
            Search: 'Search',
            And: 'And',
            Or: 'Or',
            /* Text Compare */
            Text: 'Text',
            TextContains: 'Contains',
            TextEmpty: 'Empty',
            TextStartsWith: 'Starts With',
            TextEndsWith: 'Ends With',
            TextEquals: 'Equals',
            TextNotEmpty: 'Not Empty',
            /* Number Compare */
            Number: 'Number',
            NumberEquals: 'Equals',
            NumberNotEquals: 'Not Equals',
            NumberGreaterThan: 'Greater Than',
            NumberLessThan: 'Less Than',
            NumberEmpty: 'Empty',
            NumberNotEmpty: 'Not Empty',
            /* Category List Compare */
            CategoryContains: 'Contains',
            CategoryNotContains: 'Not Contains',
        };
    }
    TableIntl.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ TableIntl.ɵprov = i0.ɵɵdefineInjectable({ factory: function TableIntl_Factory() { return new TableIntl(); }, token: TableIntl, providedIn: "root" });
    return TableIntl;
}());
export { TableIntl };
if (false) {
    /** @type {?} */
    TableIntl.prototype.menuLabels;
    /** @type {?} */
    TableIntl.prototype.paginatorLabels;
    /** @type {?} */
    TableIntl.prototype.tableLabels;
    /** @type {?} */
    TableIntl.prototype.filterLabels;
}
