// dynamic-mat-table/esm5/lib/dynamic-mat-table/extensions/toolbar/toolbar.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-mat-table/extensions/toolbar/toolbar.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, HostListener, Input, Output } from '@angular/core';
var ToolbarComponent = /** @class */ (function () {
    function ToolbarComponent(_el) {
        this._el = _el;
        this.itemList = [];
        this.color = null;
        this.direction = 'ltr';
        this.actionClick = new EventEmitter();
        this.normalItemList = [];
        this.floatItemList = [];
    }
    /**
     * @param {?} e
     * @return {?}
     */
    ToolbarComponent.prototype.onKeyup = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    ToolbarComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.itemList && this.itemList) {
            this.normalItemList = this.itemList.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.float === undefined || item.float == false; }));
            this.floatItemList = this.itemList.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return (item === null || item === void 0 ? void 0 : item.float) == true; }));
        }
    };
    /**
     * @return {?}
     */
    ToolbarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} item
     * @return {?}
     */
    ToolbarComponent.prototype.item_onClick = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.actionClick.emit(item);
    };
    ToolbarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dm-toolbar',
                    template: "<div class=\"toolbar\">\n  <div *ngFor=\"let item of normalItemList\" class=\"item\">\n    <a *ngIf=\"item?.splitter\" class=\"splitter\"></a>\n    <div  \n      matRipple\n      [matRippleColor]=\"color\"\n      class=\"toolbar-item\"\n      (click)=\"item_onClick(item)\"    \n      [matTooltip]=\"item?.tooltip\" \n      matTooltipClass=\"custom-tooltip\">  \n      <mat-icon *ngIf=\"item.matIcon\" [style.color]=\"item?.matIconColor\">{{\n        item.matIcon\n      }}</mat-icon>\n    </div>\n  </div>\n  <div [class.float-right]=\"direction === 'ltr'\" [class.float-left]=\"direction === 'rtl'\">\n    <div *ngFor=\"let item of floatItemList\">\n      <a *ngIf=\"item?.splitter\" class=\"splitter\"></a>\n      <div  \n        matRipple\n        [matRippleColor]=\"color\"\n        class=\"toolbar-item\"\n        (click)=\"item_onClick(item)\"    \n        [matTooltip]=\"item?.tooltip\" \n        matTooltipClass=\"custom-tooltip\">  \n        <mat-icon *ngIf=\"item.matIcon\" [style.color]=\"item?.matIconColor\">\n          {{item.matIcon}}</mat-icon>\n      </div>\n    </div>\n  </div>\n</div> \n<!-- {{ _el?.nativeElement?.scrollWidth}} \n{{ _el?.nativeElement?.offsetWidth}}  -->",
                    styles: [":host{background-image:linear-gradient(#f5f5f5,#e4e3e3,#fafafa);min-height:42px;border-radius:8px;border:1px solid #adadad;box-sizing:border-box;overflow:hidden}.toolbar{display:flex;align-items:center;flex-direction:row;align-self:stretch;height:calc(100% - 8px);margin-top:3px}.item{height:100%}.toolbar-item{width:34px;height:100%;border-radius:8px;background-color:#ffffffa6;border:1px solid #b0b0b0;margin-right:8px;-webkit-user-select:none;-moz-user-select:none;user-select:none;cursor:pointer}.toolbar-item:hover{background-color:#fff!important}.toolbar-item mat-icon{font-size:32px;height:100%;width:100%;text-align:center}.splitter{background-color:#b0b0b0;min-height:100%;min-width:2px;margin-right:8px;float:left;-webkit-user-select:none;-moz-user-select:none;user-select:none}::ng-deep .custom-tooltip{background-color:#fff!important;color:#000!important;border:1px solid #adadad;margin-top:-1px!important;margin-left:5px!important;text-align:center!important;font-family:IRANSans;font-weight:700;font-size:12px;min-width:100px}.float-right{margin-left:auto;display:flex;height:100%}.float-right .toolbar-item{margin-right:4px}.float-left{margin-right:auto;display:flex;height:100%}.float-left .toolbar-item{margin-left:4px}"]
                }] }
    ];
    /** @nocollapse */
    ToolbarComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    ToolbarComponent.propDecorators = {
        itemList: [{ type: Input }],
        color: [{ type: Input }],
        direction: [{ type: Input }],
        actionClick: [{ type: Output }],
        onKeyup: [{ type: HostListener, args: ['document:keyup', ['$event'],] }]
    };
    return ToolbarComponent;
}());
export { ToolbarComponent };
if (false) {
    /** @type {?} */
    ToolbarComponent.prototype.itemList;
    /** @type {?} */
    ToolbarComponent.prototype.color;
    /** @type {?} */
    ToolbarComponent.prototype.direction;
    /** @type {?} */
    ToolbarComponent.prototype.actionClick;
    /** @type {?} */
    ToolbarComponent.prototype.normalItemList;
    /** @type {?} */
    ToolbarComponent.prototype.floatItemList;
    /** @type {?} */
    ToolbarComponent.prototype._el;
}
/**
 * @record
 */
export function ToolbarItem() { }
if (false) {
    /** @type {?} */
    ToolbarItem.prototype.id;
    /** @type {?|undefined} */
    ToolbarItem.prototype.name;
    /** @type {?|undefined} */
    ToolbarItem.prototype.tooltip;
    /** @type {?|undefined} */
    ToolbarItem.prototype.matIcon;
    /** @type {?|undefined} */
    ToolbarItem.prototype.matIconColor;
    /** @type {?|undefined} */
    ToolbarItem.prototype.svgIcon;
    /** @type {?|undefined} */
    ToolbarItem.prototype.splitter;
    /** @type {?|undefined} */
    ToolbarItem.prototype.disable;
    /** @type {?|undefined} */
    ToolbarItem.prototype.float;
}
